internal/regexpchecker/regexpchecker.go

```diff
@@ -0,0 +1,1351 @@
+	if v.surrogateState != nil {
+		low := v.surrogateState.lowSurrogate
+		size := v.surrogateState.utf8Size
+		v.surrogateState = nil
+		v.pos += size
+		// Return the low surrogate encoded as UTF-16BE
+		return encodeSurrogate(low)
+	}
+
+	// Decode the next UTF-8 character from the source
+	r, s := v.charAndSize()
+
+	if v.anyUnicodeMode || r < supplementaryMin {
+		// In Unicode mode, or for BMP characters, consume and return normally
+		v.pos += s
+		return v.text[v.pos-s : v.pos]
+	}
+
+	// In non-Unicode mode with a supplementary character (>= U+10000):
+	// JavaScript represents these as surrogate pairs (two UTF-16 code units).
+	// Return the high surrogate now and save the low surrogate for the next call.
+	high, low := splitToSurrogatePair(r)
+
+	v.surrogateState = &surrogatePairState{
+		lowSurrogate: low,
+		utf8Size:     s,
+	}
+
+	return encodeSurrogate(high)
+}
+
+// ClassRanges ::= ClassAtom ('-' ClassAtom)?
+// Scans character class content like [a-z] or [^0-9].
+// Follows ECMAScript regexp grammar
+func (v *regExpValidator) scanClassRanges() {
+	isNegated := v.charAtOffset(0) == '^'
+	if isNegated {
+		v.pos++
+	}
+	oldIsCharacterComplement := v.isCharacterComplement
+	v.isCharacterComplement = isNegated
+	defer func() {
+		v.isCharacterComplement = oldIsCharacterComplement
+	}()
+	for {
+		ch := v.charAtOffset(0)
+		if v.isClassContentExit(ch) {
+			return
+		}
+		atomStart := v.pos
+		atom := v.scanClassAtom()
+		if v.charAtOffset(0) == '-' && v.charAtOffset(1) != ']' {
+			v.pos++
+			if v.isClassContentExit(v.charAtOffset(0)) {
+				return
+			}
+			// Check if min side of range is a character class escape
+			if atom == "" && v.anyUnicodeModeOrNonAnnexB {
+				v.error(diagnostics.A_character_class_range_must_not_be_bounded_by_another_character_class, atomStart, v.pos-1-atomStart)
+			}
+			rangeEndStart := v.pos
+			rangeEnd := v.scanClassAtom()
+			// Check if max side of range is a character class escape
+			if rangeEnd == "" && v.anyUnicodeModeOrNonAnnexB {
+				v.error(diagnostics.A_character_class_range_must_not_be_bounded_by_another_character_class, rangeEndStart, v.pos-rangeEndStart)
+			}
+			// Check range order
+			if atom != "" && rangeEnd != "" {
+				minCodePoint := decodeCodePoint(atom)
+				maxCodePoint := decodeCodePoint(rangeEnd)
+
+				// Get the expected sizes (in UTF-16 code units)
+				minExpectedSize := charSize(minCodePoint)
+				maxExpectedSize := charSize(maxCodePoint)
+
+				// Check if both are "complete" characters in JavaScript's UTF-16 representation.
+				// A character is complete if its UTF-16 length matches the expected size.
+				// In JavaScript, string.length returns the UTF-16 code unit count.
+				minUTF16Length := utf16Length(atom)
+				maxUTF16Length := utf16Length(rangeEnd)
+
+				minIsComplete := minUTF16Length == minExpectedSize
+				maxIsComplete := maxUTF16Length == maxExpectedSize
+
+				if minIsComplete && maxIsComplete && minCodePoint > maxCodePoint {
```

Comment:
The original intention is merely to determine whether an atom has only a single character (codepoint) for a valid order comparison. I believe `utf8.RuneCountInString(atom) == 1` is already sufficient (and is actually an equivalent), but I might be wrong. (In the original TypeScript code, I could have written `[...atom].length === 1` – comparing by `charSize` is purely for avoiding array allocation.)